src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/BulkCopy/SqlBulkCopy.Async.cs

```diff
@@ -0,0 +1,1108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Data;
+using System.Data.Common;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+
+namespace Microsoft.Data.SqlClient
+{
+    public partial class SqlBulkCopy
+    {
+        // Newly added member variables for Async modification, m = member variable to bcp.
+        private int _savedBatchSize = 0; // Save the batchsize so that changes are not affected unexpectedly.
+        private bool _hasMoreRowToCopy = false;
+
+        // we can remove this variable as it is not being used anymore.
+        private bool _isAsyncBulkCopy = false;
+        private bool _isBulkCopyingInProgress = false;
+        private SqlInternalConnectionTds.SyncAsyncLock _parserLock = null;
+
+        /// <include file='../../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlBulkCopy.xml' path='docs/members[@name="SqlBulkCopy"]/WriteToServerAsync[@name="DataRowParameter"]/*'/>
+        public async Task WriteToServerAsync(DataRow[] rows) => await WriteToServerAsync(rows, CancellationToken.None);
```

Comment:
Why are we awaiting these methods?

Reply:
Unrelated comment: it is possible to remove the async and await keywords here, and just return the Task directly. This removes an async state machine and helps performance.

This trick can be done anywhere where the await call is the last line in a method.